### PR TITLE
Spells on Invisible Creatures ( Mostly Scripted )

### PR DESCRIPTION
**Pull request by aaron02**
Added the Ability to Cast Spells near Creatures / Gos needed for certain
quests where you need to cast a spell near a certain place

### CODE REVIEW

File: src/world/Spell/Spell.cpp [hunk 1]
```diff
@@ -4062,16 +4062,46 @@ uint8 Spell::CanCast(bool tolerate)
...
-                if (target->IsGameObject() && !m_target_constraint->HasGameobject(target->GetEntry()))
+                if (target->IsGameObject() && !m_target_constraint->HasGameobject(target->GetEntry()) && !m_target_constraint->IsFocused(target->GetEntry()))
+                    return SPELL_FAILED_BAD_TARGETS;
+
+                bool Target = false;
+                Creature* pCreature = nullptr;
+                size_t creatures = m_target_constraint->GetCreatures().size();
+
+                // Spells for Invisibl Creatures and or Gameobjects ( Casting Spells Near them )
+                for (size_t i = 0; i < creatures; ++i)
```
Comment (Evairfairy) [inline src/world/Spell/Spell.cpp]: Do not use braceless _for_ loops

File: src/world/Spell/Spell.cpp [hunk 2]
```diff
@@ -4062,16 +4062,46 @@ uint8 Spell::CanCast(bool tolerate)
...
+                if (!m_target_constraint->IsFocused(m_target_constraint->GetCreatures()[i]))
+                {
+                    Creature* pCreature = m_caster->GetMapMgr()->GetInterface()->GetCreatureNearestCoords(m_caster->GetPositionX(), m_caster->GetPositionY(), m_caster->GetPositionZ(), m_target_constraint->GetCreatures()[i]);
```
Comment (Evairfairy) [inline src/world/Spell/Spell.cpp]: Redefinition of pCreature from line 4074

File: src/world/Spell/Spell.cpp [hunk 3]
```diff
@@ -4062,16 +4062,46 @@ uint8 Spell::CanCast(bool tolerate)
...
+
+                    if( pCreature)
```
Comment (Evairfairy) [inline src/world/Spell/Spell.cpp]: Do not use braceless if statements containing blocks

File: src/world/Spell/Spell.cpp [hunk 4]
```diff
@@ -4062,16 +4062,46 @@ uint8 Spell::CanCast(bool tolerate)
...
+                        {
+                            target = pCreature;
+                            Target = true;
+                        }
+                }
+
+                GameObject* pGameobject = nullptr;
+                size_t gameobjects = m_target_constraint->GetGameobjects().size();
+
+                for (size_t i = 0; i < gameobjects; ++i)
```
Comment (Evairfairy) [inline src/world/Spell/Spell.cpp]: Do not use braceless for loops

File: src/world/Spell/Spell.cpp [hunk 5]
```diff
@@ -4062,16 +4062,46 @@ uint8 Spell::CanCast(bool tolerate)
...
+                if (!m_target_constraint->IsFocused(m_target_constraint->GetGameobjects()[i]))
+                {
+                    pGameobject = m_caster->GetMapMgr()->GetInterface()->GetGameObjectNearestCoords(m_caster->GetPositionX(), m_caster->GetPositionY(), m_caster->GetPositionZ(), m_target_constraint->GetGameobjects()[i]);
+
+                    if (pGameobject)
```
Comment (Evairfairy) [inline src/world/Spell/Spell.cpp]: Do not use braceless if statements containing blocks

File: src/world/Spell/Spell.cpp [hunk 6]
```diff
@@ -4062,16 +4062,46 @@ uint8 Spell::CanCast(bool tolerate)
...
-                if (target->IsCreature() && !m_target_constraint->HasCreature(target->GetEntry()))
+                if (target->IsCreature() && !m_target_constraint->HasCreature(target->GetEntry()) && !m_target_constraint->IsFocused(target->GetEntry()))
                     return SPELL_FAILED_BAD_TARGETS;
 
                 // target is the wrong GO :/
-                if (target->IsGameObject() && !m_target_constraint->HasGameobject(target->GetEntry()))
+                if (target->IsGameObject() && !m_target_constraint->HasGameobject(target->GetEntry()) && !m_target_constraint->IsFocused(target->GetEntry()))
+                    return SPELL_FAILED_BAD_TARGETS;
+
+                bool Target = false;
```
Comment (Evairfairy) [inline src/world/Spell/Spell.cpp]: Inconsistent naming - should be _target_ or similar

**Comment by DankoDJ** [comment]
Do you actually work on a quest script using the new possibilities?

**Comment by aaron02** [comment]
Yes ;)

File: src/scripts/QuestScripts/Quest_BoreanTundra.cpp [hunk 7]
```diff
@@ -640,6 +640,63 @@ class GuardMitchGossip : public GossipScript
         }
 };
 
+bool PlaceCart(uint32 i, Spell* pSpell)
+{
+
```
Comment (DankoDJ) [inline src/scripts/QuestScripts/Quest_BoreanTundra.cpp]: single empty line

File: src/scripts/QuestScripts/Quest_BoreanTundra.cpp [hunk 8]
```diff
@@ -640,6 +640,63 @@ class GuardMitchGossip : public GossipScript
...
+    Player* pPlayer = pSpell->p_caster;
+    if (pPlayer == nullptr)
+        return true;
+
+    Creature* pUnit = pSpell->GetTargetConstraintCreature();
```
Comment (DankoDJ) [inline src/scripts/QuestScripts/Quest_BoreanTundra.cpp]: why pUnit? The type is creature :)

Comment (aaron02) [inline src/scripts/QuestScripts/Quest_BoreanTundra.cpp]: The moment when you know that you fucked up :P

File: src/world/Spell/Spell.cpp [hunk 9]
```diff
@@ -312,6 +312,8 @@ Spell::Spell(Object* Caster, SpellEntry* info, bool triggered, Aura* aur)
     gameObjTarget = NULL;
     playerTarget = NULL;
     corpseTarget = NULL;
+    targetConstraintCreature = NULL;
```
Comment (DankoDJ) [inline src/world/Spell/Spell.cpp]: NULL hmmmm.... isn't it a pointer? So initialize it as nullptr.

**Comment by aaron02** [comment]
I change it ;)

**Comment by DankoDJ** [comment]
Guessed SQL update:
UPDATE `spelltargetconstraints` SET `TargetType`=2 WHERE `SpellID`=46797;
?

**Comment by aaron02** [comment]
Sure :)

**Pull request merged**